homeassistant/components/labs/__init__.py

```diff
@@ -0,0 +1,218 @@
+"""The Home Assistant Labs integration.
+
+This integration provides preview features that can be toggled on/off by users.
+Integrations can register lab features in their manifest.json which will appear
+in the Home Assistant Labs UI for users to enable or disable.
+"""
+
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.components import websocket_api
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.generated.labs import LABS_FEATURES
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.storage import Store
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.loader import async_get_custom_components
+
+from .const import (
+    DOMAIN,
+    EVENT_LABS_UPDATED,
+    LABS_DATA,
+    STORAGE_KEY,
+    STORAGE_VERSION,
+    EventLabsUpdatedData,
+    LabFeature,
+    LabsData,
+    LabsStoreData,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)
+
+__all__ = [
+    "DOMAIN",
+    "EVENT_LABS_UPDATED",
+    "EventLabsUpdatedData",
+    "async_is_preview_feature_enabled",
+]
+
+
+async def _async_scan_all_lab_features(hass: HomeAssistant) -> dict[str, LabFeature]:
+    """Scan ALL available integrations for lab features (loaded or not)."""
+    features: dict[str, LabFeature] = {}
+
+    # Load pre-generated built-in labs features (already includes all data)
+    for domain, domain_features in LABS_FEATURES.items():
+        for feature_id, feature_data in domain_features.items():
+            feature = LabFeature(
+                domain=domain,
+                feature=feature_id,
+                feedback_url=feature_data.get("feedback_url"),
+                learn_more_url=feature_data.get("learn_more_url"),
+                report_issue_url=feature_data.get("report_issue_url"),
+            )
+            features[feature.full_key] = feature
+
+    # Scan custom components
+    custom_integrations = await async_get_custom_components(hass)
+    _LOGGER.debug(
+        "Loaded %d built-in + scanning %d custom integrations for lab features",
+        len(features),
+        len(custom_integrations),
+    )
+
+    for integration in custom_integrations.values():
+        if not (labs_features := integration.labs_features):
+            continue
+
+        # labs_features is a dict[str, dict[str, str]]
+        for feature_key, feature_data in labs_features.items():
```

Comment:
Maybe have a small helper function that iterates and populates the features dict? We're doing that twice now, once for core integrations and once for custom integrations.